### PR TITLE
Add FlatFSAdapter

### PR DESCRIPTION
**Pull request by adamjsawicki**
Implements a FlatFSAdapter based on https://github.com/ipfs/go-ds-flatfs (@Kubuxu)

It is an implementation of a KeyTransform Adapter, as noted by @alexander255 [here](https://github.com/ipfs/py-datastore/issues/10#issuecomment-579332311).

A few notes:
* Unless I'm missing something about the implementation, involving a `SHARDING` file at all seems superfluous. It seems more in line with the other adapters, especially `NestedPath` that the `FlatFSAdapter` would take as arguments `sharding_function_name: str` and `key_length: int` that could have defaults as opposed to doing any type of IO and making the user adhere to what seems to be an arbitrary spec.
* Not sure why the `default_prefix` is what it is
* Not sure why we have versioning at all if we only accept `v1`

A few questions:
* Should we ship with a `SHARDING` file. If so, where should it be kept?
* What is the best way to test this? Is there a good way to mock what is in `SHARDING`?
    * On this note, I suppose we could pull out `_parse_sharding_function` a helper function `_get_sharding_function_string` that could be mocked. Thoughts?

### CODE REVIEW

File: datastore/adapter/keytransform.py [hunk 1]
```diff
@@ -290,3 +295,161 @@ class BinaryNestedPathAdapter(_NestedPathAdapter[datastore.abc.BinaryDatastore],
...
+		>>> import datastore
+		>>>
+		>>> ds = datastore.DictDatastore()
+		>>> np = datastore.FlatFSDatastore(ds)
+		>>>
+		>>> np.put(datastore.Key('/abcdefghijk'), 1)
+		>>> np.get(datastore.Key('/abcdefghijk'))
+		1
+		>>> ds.get(datastore.Key('/abcdefghijk'))
+		None
```
Comment (ntninja) [inline datastore/adapter/keytransform.py]: Note: Actually it would raise KeyError here

File: datastore/adapter/keytransform.py [hunk 2]
```diff
@@ -290,3 +295,161 @@ class BinaryNestedPathAdapter(_NestedPathAdapter[datastore.abc.BinaryDatastore],
...
+		    raise Exception(f"{self.sharding_file} file does not exist")
+
+		with open(path, "r") as f:
+		    sharding_func: str = f.read()
+
+		# Ensure not empty
+		if not sharding_func:
+		    raise Exception("Empty shard identifier file.")
+
+		sharding_func = sharding_func.strip()
```
Comment (ntninja) [inline datastore/adapter/keytransform.py:477]: You need to read this from the backend datastore using something like

```py
sharding_func: str
try:
    # Try reading existing sharding func used
    sharding_func = self.child_datastore.get_all(self.sharding_key)
except KeyError:
    # Use the given default sharding func, remembering it for subsequent use of this repo
    sharding_func = self.default_sharding_func
    await self.child_datastore.put(self.sharding_key, sharding_func)
```

Where `self.default_sharding_func` is a parameter given in `__init__`. (Possibly with a default value.)

File: datastore/adapter/keytransform.py [hunk 3]
```diff
@@ -290,3 +295,161 @@ class BinaryNestedPathAdapter(_NestedPathAdapter[datastore.abc.BinaryDatastore],
...
+	@staticmethod
+	def _suffix(key: datastore.Key, length: int) -> datastore.Key:
+		return datastore.Key(str(key).rjust(length, "_")[:length])
+
+	@staticmethod
+	def _next_to_last(key: datastore.Key, length: int) -> datastore.Key:
+		return datastore.Key(str(key).rjust(length + 1, "_")[:length])
+
+	def __init__(self, *args,
+	             sharding_file: str = None,
```
Comment (ntninja) [inline datastore/adapter/keytransform.py]: `sharding_file` is a datastore key, not an actual file path. Change this (and all related properties) to `sharding_key: datastore.Key` instead.

File: datastore/adapter/keytransform.py [hunk 4]
```diff
@@ -290,3 +295,161 @@ class BinaryNestedPathAdapter(_NestedPathAdapter[datastore.abc.BinaryDatastore],
...
+		if not sharding_func:
+			raise Exception(f"Prefix ({self.prefix}) was not present in {sharding_func}")
+		
+		parts: list = sharding_func.split("/")
+
+		if len(parts) != 3:
+			raise Exception(
+				f"invalid shard identifier: {sharding_func}.\n"
+				f"Expecting form: {self.prefix}/version/function_name/key_length"
+			)
```
Comment (ntninja) [inline datastore/adapter/keytransform.py]: :+1:

File: datastore/adapter/keytransform.py [hunk 5]
```diff
@@ -290,3 +295,161 @@ class BinaryNestedPathAdapter(_NestedPathAdapter[datastore.abc.BinaryDatastore],
...
+
+		if version != "v1":
+			raise Exception(f"Expected 'v1' for version string got: {version}\n")
+
+		try:
+			length = int(length)
+		except ValueError:
+			raise Exception(f"Invalid parameter: {length}. Should be integer representing `key_length`.")
+
+		__funcs: dict = {
```
Comment (ntninja) [inline datastore/adapter/keytransform.py]: Nit: Don't use double underscores in Python (they can invoke magic sometimes), just name this `funcs` instead.

File: datastore/adapter/keytransform.py [hunk 6]
```diff
@@ -290,3 +295,161 @@ class BinaryNestedPathAdapter(_NestedPathAdapter[datastore.abc.BinaryDatastore],
...
+			"prefix": self._prefix,
+			"suffix": self._suffix,
+			"next-to-last": self._next_to_last
+		}
+
+		try:
+			_func = __funcs[function_name]
```
Comment (ntninja) [inline datastore/adapter/keytransform.py]: No need for this underscore either

**Review by ntninja — CHANGES_REQUESTED** [review]

**Comment by ntninja** [comment]
> Unless I'm missing something about the implementation, involving a SHARDING file at all seems superfluous. It seems more in line with the other adapters, especially NestedPath that the FlatFSAdapter would take as arguments sharding_function_name: str and key_length: int that could have defaults as opposed to doing any type of IO and making the user adhere to what seems to be an arbitrary spec.

Checking the value of `SHARDING` is hugely important so that we don't corrupt existing datastores that don't use our preferred sharding function. Compatibility with the existing spec is an important goal here IMHO.

> Not sure why the default_prefix is what it is

Nor do it, it's just the way it is. :wink: 

> Not sure why we have versioning at all if we only accept v1

Just a safety measure in case the format of the SHARDING file or other stuff where to every change significantly.

> Should we ship with a SHARDING file. If so, where should it be kept?

I implicitly answered this in my review already, be just write a new `SHARDING` key/file with our preferred sharding function if there is none already.

> What is the best way to test this? Is there a good way to mock what is in SHARDING?

Some ideas:

  * Copy some (small) real-world repositories from go-ipfs to the test directory and query those.
  * Layer this adapter on top of `DictDatastore` (like in your doctest example) and set the value of the `SHARDING` key on that before initializing this adapter.

File: datastore/adapter/keytransform.py (hunk 1 above)
Comment (adamjsawicki) [reply datastore/adapter/keytransform.py]: Blindly copied from the outdated ones in `NestedPath`. Would appreciate guidance regarding my [comment](https://github.com/ipfs/py-datastore/issues/16#issuecomment-581173361) on #16 and then will fix up.

Comment (ntninja) [reply datastore/adapter/keytransform.py]: You're right, they're all wrong either way! :stuck_out_tongue: 

Unfortunately, for now `trio.run(lambda: …)` will be the best we can do however.

File: datastore/adapter/keytransform.py (hunk 2 above)
Comment (adamjsawicki) [reply datastore/adapter/keytransform.py:477]: Using `put` will require that we `await`, making the `_parse_sharding_func` method `async`, requiring that we `await` in `__init__`, making `__init__` `async`, which will throw:
```
TypeError: __init__() should return None, not 'coroutine'
```

Comment (adamjsawicki) [reply datastore/adapter/keytransform.py:477]: I think I have a gap in my understanding: To have access to `self. child_datastore` we need to inherit from `Datastore.ADAPTER_T`, where it gets [initialized](https://github.com/ipfs/py-datastore/blob/e566d40a8ca81d8628147e255fe7830b5f928a43/datastore/core/objectstore.py#L348), but `_Adapter` currently inherits from `typing.Generic[DS]`.

Comment (ntninja) [reply datastore/adapter/keytransform.py:477]: `_Adapter` is not used by itself, but always a mixin class in `BinaryAdapter` and `ObjectAdapter`. These two types are conceptually very similar, but still strictly distinct types (they “look” very similar in terms of API, but represent different things). Since object and binary adapters are so similar API-wise when can actually implement both of these using just one piece of code, however the *meaning* of this code will vary depending on which context it is used it.

**Maybe this motivational story makes it clearer**: You, as a developer, want to expose a new datastore adapter (which you are doing) and to do that you'll need to implement two similar interfaces: `ObjectAdapter` and `BinaryAdapter`. They are very similar but not quite the same. So similar in fact, that if you were them implement them separately anyway, you would end up with what is essentially code duplication. You could move the identical code to separate functions, but now all your methods are left to do is calling those functions, and again the code that does this looks identical in both implementations. So you think: “Maybe I can somehow implement *both* of these interfaces in the *same* class?” So you create a new class and move all code there, but you cannot decide whether this should now be an `ObjectAdapter` or a `BinaryAdapter` – It's kinda both, or rather it *could* be both, but it's not both *at the same time*. You'll still need expose two separate classes – one for each type – they should just share the same method implementations. Turns out Python has a mechanism for exactly that: Multiple Inheritance! By inheriting from *both* the base class you'd like to implement and the extra class containing the just the actual method code, you end up with sharing the *code* between the two classes while still keeping them separate from a type perspective.

In other words: Think about the `_Adapter` class of just being a *container* for storing the methods you'd like to be visible in both the *ObjectAdapter* and the *BinaryAdapter* implementation.

And ignore that we are inheriting from `typing.Generic[DS]` there, that's just something make the types used during type checking (`tox -e typeck`) more accurate.

**Comment by ntninja** [comment]
Regarding `await`: You just stumbled on the joys of [“two-colored“ functions](https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/) there… 

To solve that it's easiest to just stuff the entire `await` startup code into a new construction function and be done with it. Something like:

```py
T = TypeVar('T', bound="_Adapter")
class _Adapter(…):
    …
    @classmethod
    async def create(cls: typing.Type[T], child_datastore: DS, *, default_sharding_fn: str = …, sharding_key: datastore.Key = None, prefix: str = None,**kwargs) -> T:
       sharding_fn = await <the-sharding-file-stuff>
       return cls(child_datastore, *, _sharding_fn=sharding_fn, **kwargs)
```

BTW: I don't think `prefix` should be configurable as a parameter here, better to make it a class level constant.

File: datastore/adapter/keytransform.py (hunk 2 above)
Comment (adamjsawicki) [reply datastore/adapter/keytransform.py:477]: Thank you for a wonderful explanation of Mixins. Makes a ton of sense.

**Comment by adamjsawicki** [comment]
Thanks for sending me down that rabbit hole 😅 

For the `create` function, do you see an issue with having one of the `Adapters` use a factory pattern while none of the others do? Or do you want all of the `Adapters` to conform to that pattern?

Additionally, in regards to the `child_datastore`, we still need access to it in the _initialization_ of `FlatFSAdapter`, meaning that the top level object (i.e. one that inherits from both this `Adapter` and a `Datastore`) that would ultimately have access to a `child_datastore` does not exist yet, still barring our access.

**Comment by ntninja** [comment]
> For the create function, do you see an issue with having one of the Adapters use a factory pattern while none of the others do? Or do you want all of the Adapters to conform to that pattern?

No, turns out at least the FileSystemDatastore needs one as well as it currently calls `mkdir` synchronously and I'd very surprised if some of the other backends (like leveldb) *didn't* need one at some point. Based on this, I'll actually add a default async `.create()` method to all datastores that just delegates to the synchronous `__init__` soon and datastores that don't support synchronous initialization can then just reject that mode of creation.

> Additionally, in regards to the child_datastore, we still need access to it in the initialization of FlatFSAdapter, meaning that the top level object (i.e. one that inherits from both this Adapter and a Datastore) that would ultimately have access to a child_datastore does not exist yet, still barring our access.

In the `create` example I posted it was already available as part of the somewhat opaque `args` parameter – I've slightly updated the example to make that clearer. Bottom line is: Since the value is forwarded as an argument to the the constructor of `DatastoreAdapter`, if had to be present, one way or another, at this level as well; it wasn't very obvious *where* it was.

**Comment by adamjsawicki** [comment]
Made the changes suggested. Thanks for all the help thus far!

I can add the default async `.create()` method  and some documentation around it in another PR/appended to this one if you'd like.

File: datastore/adapter/keytransform.py [hunk 7]
```diff
@@ -338,3 +341,204 @@ class ObjectNestedPathAdapter(
...
+		Exception
+			Prefix was not present in sharding file
+		Exception
+			Sharding file was in incorrect format
+		Exception
+			Expect `v1` format
+		Exception
+			`key_length` shouold be an integer
+		Exception
+			`function_name` was not `prefix`, `suffix`, or `next-to-last`
```
Comment (ntninja) [inline datastore/adapter/keytransform.py:454]: Doing `raise Exception("<some text>")` is an anti-pattern in Python (unlike JavaScript where the exception type is mostly useless anyways). Please add an exception hierarchy like the following for this:

```py
class InitializationError(RuntimeError):
	"""Raised when there was an error with opening this datastore adapter."""
	datastore: T
	
	def __init__(self, datastore: T, *args):
		self.datastore = datastore
		super().__init__(*args)

# Use for other errors
class ShardingStateError(InitializationError):
	"""Raised when the existing sharding entry could not be read."""
	key: datastore.Key
	
	def __init__(self, datastore: T, key: datastore.Key, *args):
		self.key = key
		super().__init__(datastore, *args)

# Use for: Empty shard identifier file.
class ShardingStateEmptyError(ShardingStateError):
	"""Raised when the existing sharding entry is empty."""

# Use for: Expect `v1` format & `function_name` was not `prefix`, `suffix`, or `next-to-last`
class ShardingStateUnsupportedError(ShardingStateError):
	"""Raised when the existing sharding entry uses an unsupported version number or sharding function."""
```

(The comment lines are just for your orientation.) Add this to the file top, add the relevant entries to `__all__` and update all raise statements to `raise <ErrorType>(self, sharding_key, "<some text>")`. This way consumers of this API can more easily distinguish between the kinds of errors raised (using `try: …; except <ErrorType> as exc: …`). Also some extra error data is included to help callers understand in which context the error occurred.

File: datastore/adapter/keytransform.py [hunk 8]
```diff
@@ -338,3 +341,204 @@ class ObjectNestedPathAdapter(
...
+
+		prefix: str = cls.prefix  # type: ignore[attr-defined] # noqa: F821
+
+		sharding_func: str
+		need_to_store_sharding_func: bool = False
+		try:
+			# Try reading existing sharding func used
+			sharding_func = next(   # type: ignore[assignment] # noqa: F821
+				iter(await child_datastore.get_all(sharding_key))
+			)
```
Comment (ntninja) [inline datastore/adapter/keytransform.py:467]: This looks like it will return a binary string / bytes object when `self` is `BinaryAdapter`, but you need a Unicode string so you'll need to call `decode` somewhere. I propose the following to support for object and binary datastores:

```py
if isinstance(self, datastore.abc.BinaryAdapter):
	# Decode using latin-1 to get a 1:1 binary to Unicode mapping that cannot fail
	sharding_func = (await child_datastore.get_all(sharding_key)).decode("latin-1")
else:
	# Expect result from object datastore be a Unicode string
	sharding_func = typing.cast(str, await child_datastore.get_all(sharding_key))
	assert isinstance(sharding_func, str)
```

File: datastore/adapter/keytransform.py [hunk 9]
```diff
@@ -338,3 +341,204 @@ class ObjectNestedPathAdapter(
...
+		except KeyError:
+			# Use the given default sharding func
+			sharding_func = default_sharding_func
+			need_to_store_sharding_func = True
+
+		# Ensure not empty
+		if not sharding_func:
+			raise Exception("Empty shard identifier file.")
+
+		sharding_func = sharding_func.strip()
```
Comment (ntninja) [inline datastore/adapter/keytransform.py:477]: Place this line above the above exception as `bool(" ") is True` (not `False`)

File: datastore/adapter/keytransform.py [hunk 10]
```diff
@@ -338,3 +341,204 @@ class ObjectNestedPathAdapter(
...
+
+		# Returns a 3-tuple containing the part before the separator,
+		# the separator itself, and the part after the separator
+		_, _, sharding_fn = sharding_func.partition(prefix)
+
+		# Ensure proper format
+		if not sharding_fn:
+			raise Exception(f"Prefix ({prefix}) was not present in {sharding_func}")
+
+		parts: list = sharding_fn.split("/")
```
Comment (ntninja) [inline datastore/adapter/keytransform.py:487]: Use the type `tying.List[str]` here to prevent type erasure.

File: datastore/adapter/keytransform.py [hunk 11]
```diff
@@ -338,3 +341,204 @@ class ObjectNestedPathAdapter(
...
+		try:
+			length = int(length)
+		except ValueError:
+			raise Exception(f"Invalid parameter: {length}. Should be integer representing `key_length`.")
+
+		funcs: dict = {
+			"prefix": cls._prefix,  # type: ignore[attr-defined] # noqa: F821
+			"suffix": cls._suffix,  # type: ignore[attr-defined] # noqa: F821
+			"next-to-last": cls._next_to_last  # type: ignore[attr-defined] # noqa: F821
+		}
```
Comment (ntninja) [inline datastore/adapter/keytransform.py:509]: Make this a class variable and change the type to `typing.Dict[str, typing.Callable[[datastore.Key, int], datastore.Key]]` for a proper typing signature.

File: datastore/adapter/keytransform.py [hunk 12]
```diff
@@ -338,3 +341,204 @@ class ObjectNestedPathAdapter(
...
+
+		try:
+			func = funcs[function_name]
+		except Exception:
```
Comment (ntninja) [inline datastore/adapter/keytransform.py:513]: Just like above, `except Exception:` is an antipattern in Python unless you really mean “catch all non-critical errors”. In this case you only want to catch `KeyError` (I presume), so write that instead.

Additionally, if you write something like `except …: raise <ErrorType>(…)` in Python, the interpreter will assume that your code crashed while trying to handle the previous exception and will show this in the backtrace. If you are instead raising an exception because you caught another write `except … as exc: raise <ErrorType>(…) from exc` instead so that the interpret can show proper explanations of what happened in stack traces.

File: datastore/adapter/keytransform.py [hunk 13]
```diff
@@ -338,3 +341,204 @@ class ObjectNestedPathAdapter(
...
+	Implementation Notes:
+		* The `default_prefix` is "/repo/flatfs/shard/"
+		* The only version supported is "v1"
+		* function_name must be `prefix`, `suffix`, or `next-to-last`
+
+	"""
+
+	prefix: str = "/repo/flatfs/shard/"
+	_default_sharding_key: datastore.Key = datastore.Key("SHARDING")
+	_default_sharding_func: str = "v1/prefix/2"
```
Comment (ntninja) [inline datastore/adapter/keytransform.py:362]: Use `"v1/next-to-last/2"` here as that is the default go-ipfs uses.

File: datastore/adapter/keytransform.py [hunk 14]
```diff
@@ -338,3 +341,204 @@ class ObjectNestedPathAdapter(
...
+	def _suffix(key: datastore.Key, length: int) -> datastore.Key:
+		return datastore.Key(str(key).rjust(length, "_")[:length])
+
+	@staticmethod
+	def _next_to_last(key: datastore.Key, length: int) -> datastore.Key:
+		return datastore.Key(str(key).rjust(length + 1, "_")[:length])
+
+	@classmethod
+	async def _parse_sharding_function(
+			cls: typing.Type[T],
```
Comment (ntninja) [inline datastore/adapter/keytransform.py:423]: Don't use `cls: typing.Type[T],` here, it will let you drop a bunch of mypy attr-defined comments in this method.

File: datastore/adapter/keytransform.py [hunk 15]
```diff
@@ -17,11 +18,14 @@
...
 	"ObjectNestedPathAdapter",
+	
+	"BinaryFlatFSAdapter",
+	"ObjectFlatFSAdapter"
 )
 
 
-
 KEY_TRANSFORM_T = typing.Callable[[datastore.Key], datastore.Key]
+T = typing.TypeVar('T', bound="_Adapter")
```
Comment (ntninja) [inline datastore/adapter/keytransform.py:28]: The bound here should be on `"_FlatFSAdapter"`, not just `"_Adapter"`. That may also fix many mypy complains, if I'm not mistaken…

**Review by ntninja — CHANGES_REQUESTED** [review]
Found some more issues with regards to exception handling and typing unfortunately. Also there don't seem to be any tests yet. Your on the right track however! :wink:

**Comment by ntninja** [comment]
@adamjsawicki: Are you still intending on working on this?
